pontoon/insights/utils.py

```diff
@@ -0,0 +1,120 @@
+from pontoon.insights.models import LocaleInsightsSnapshot
+
+
+def get_insights(query_filters=None):
+    """Get data required by the Insights tab.
+
+    :param django.db.models.Q query_filters: filters insights by given query_filters.
+    """
+    months = sorted(
+        [aware_datetime(year, month, 1) for year, month in get_last_months(12)]
+    )
+
+    snapshots = LocaleInsightsSnapshot.objects.filter(created_at__gte=months[0])
+
+    if query_filters:
+        snapshots = snapshots.filter(query_filters)
+
+    insights = (
+        snapshots
+        # Truncate to month and add to select list
+        .annotate(month=TruncMonth("created_at"))
+        # Group By month
+        .values("month")
+        # Select the avg/sum of the grouping
+        .annotate(
+            unreviewed_suggestions_lifespan_avg=Avg("unreviewed_suggestions_lifespan")
+        )
+        .annotate(completion_avg=Avg("completion"))
+        .annotate(human_translations_sum=Sum("human_translations"))
+        .annotate(machinery_translations_sum=Sum("machinery_translations"))
+        .annotate(new_source_strings_sum=Sum("new_source_strings"))
+        .annotate(unreviewed_avg=Avg("unreviewed_strings"))
+        .annotate(peer_approved_sum=Sum("peer_approved"))
+        .annotate(self_approved_sum=Sum("self_approved"))
+        .annotate(rejected_sum=Sum("rejected"))
+        .annotate(new_suggestions_sum=Sum("new_suggestions"))
+        # Select month and values
+        .values(
+            "month",
+            "unreviewed_suggestions_lifespan_avg",
+            "completion_avg",
+            "human_translations_sum",
+            "machinery_translations_sum",
+            "new_source_strings_sum",
+            "unreviewed_avg",
+            "peer_approved_sum",
+            "self_approved_sum",
+            "rejected_sum",
+            "new_suggestions_sum",
+        )
+        .order_by("month")
+    )
+
+    dates = [convert_to_unix_time(month) for month in months]
+
+    # Active users
+    total_managers = 0
+    total_reviewers = 0
+    total_contributors = 0
+    active_managers = 0
+    active_reviewers = 0
+    active_contributors = 0
+
+    if snapshots:
+        latest_snapshot = snapshots.latest("created_at")
+
+        total_managers = latest_snapshot.total_managers
+        total_reviewers = latest_snapshot.total_reviewers
+        total_contributors = latest_snapshot.total_contributors
+
+        active_users = latest_snapshot.active_users_last_12_months
+        active_managers = active_users["managers"]
+        active_reviewers = active_users["reviewers"]
+        active_contributors = active_users["contributors"]
+
+    # Unreviewed suggestions lifespan
+    unreviewed_lifespans = [
+        x["unreviewed_suggestions_lifespan_avg"].days for x in insights
+    ]
+
+    # Translation activity
+    completion = [round(float(x["completion_avg"])) for x in insights]
```

Comment:
I can't find any place that uses  `get_insights` function. Do you have a specific use-case in mind when you think about the results of this function? I see that it returns the lists of the values instead of their aggregates (?). Also, I feel that a large portion of this function could be handled in a database query (?)

Reply:
We use the ` get_insights` function when we [render the dashboard](https://github.com/mozilla/pontoon/pull/1760#issue-534386013). Check out the reverted "[DO NOT REVIEW]" commit.

We plot some numbers (like `completion`) in a chart over time, so we need a list of items, each representing a point in chart (e.g. a month). See the [spec mockup](https://raw.githubusercontent.com/mozilla/pontoon/master/specs/0108/mockup.png). I currently use stage to test my next patch, so it might be less reliable.

I didn't find an elegant way to get to the final results via a DB query directly.

Reply:
We could make the code following the `QuerySet` more condensed:
https://gist.github.com/mathjazz/3dc4ba0fbaeb545d32897a05f12629bb